runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/WindmillStreamSender.java

```diff
@@ -0,0 +1,147 @@
+
+  private WindmillStreamSender(
+      CloudWindmillServiceV1Alpha1Stub stub,
+      GetWorkRequest getWorkRequest,
+      AtomicReference<GetWorkBudget> getWorkBudget,
+      StreamingEngineStreamFactory streamingEngineStreamFactory,
+      WorkItemReceiver workItemReceiver) {
+    this.started = new AtomicBoolean(false);
+    this.getWorkBudget = getWorkBudget;
+    this.streamingEngineThrottleTimers = StreamingEngineThrottleTimers.create();
+
+    // All streams are memoized/cached since they are expensive to create and some implementations
+    // perform side effects on construction (i.e. sending initial requests to the stream server to
+    // initiate the streaming RPC connection). Stream instances connect/reconnect internally so we
+    // can reuse the same instance through the entire lifecycle of WindmillStreamSender.
+    this.getWorkStream =
+        Suppliers.memoize(
+            () ->
+                streamingEngineStreamFactory.createGetWorkStream(
+                    stub,
+                    withRequestBudget(getWorkRequest, getWorkBudget.get()),
+                    streamingEngineThrottleTimers.getWorkThrottleTimer(),
+                    workItemReceiver));
+    this.getKeyedDataStream =
+        Suppliers.memoize(
+            () ->
+                streamingEngineStreamFactory.createGetDataStream(
+                    stub, streamingEngineThrottleTimers.getDataThrottleTimer()));
+    this.commitWorkStream =
+        Suppliers.memoize(
+            () ->
+                streamingEngineStreamFactory.createCommitWorkStream(
+                    stub, streamingEngineThrottleTimers.commitWorkThrottleTimer()));
+  }
+
+  public static WindmillStreamSender create(
+      CloudWindmillServiceV1Alpha1Stub stub,
+      GetWorkRequest getWorkRequest,
+      GetWorkBudget getWorkBudget,
+      StreamingEngineStreamFactory streamingEngineStreamFactory,
+      WorkItemReceiver workItemReceiver) {
+    return new WindmillStreamSender(
+        stub,
+        getWorkRequest,
+        new AtomicReference<>(getWorkBudget),
+        streamingEngineStreamFactory,
+        workItemReceiver);
+  }
+
+  private static GetWorkRequest withRequestBudget(GetWorkRequest request, GetWorkBudget budget) {
+    return request.toBuilder().setMaxItems(budget.items()).setMaxBytes(budget.bytes()).build();
+  }
+
+  @SuppressWarnings("ReturnValueIgnored")
+  public void startStreams() {
+    Preconditions.checkState(
+        !getWorkBudget.get().equals(GetWorkBudget.noBudget()), "Cannot GetWork with no budget.");
+    getWorkStream.get();
+    getKeyedDataStream.get();
+    commitWorkStream.get();
+    // *stream.get() is all memoized in a threadsafe manner.
+    started.compareAndSet(false, true);
+  }
+
+  public void closeAllStreams() {
+    // Supplier<Stream>.get() starts the stream which is an expensive operation as it initiates the
+    // streaming RPCs. Do not close the streams unless they have already been started.
+    if (started.get()) {
```

Comment:
this is racy if some streams are started and others not closed.  An alternative to an atomic would be to have an object you syncrhonize on here and in startStreams so close blocks if start is still ongoing.

Or you could document that start/close are not thread-safe and should be sequenced